Faster build

darwin opened the pull request:
This builds on top of https://github.com/trezor/trezor-firmware/pull/1330.

Future improvements could be to cache /nix/store in a separate persistent docker data volume or even share it with the host. 
It takes a while to initially build the docker image from Dockerfile but I'm not hit by this because docker image is built once and is properly cached by docker. This would be an issue only if someone would be developing something which would be triggering this rebuild.

tsusanka commented on build-docker.sh:
```diff
@@ -1,27 +1,104 @@
+ALPINE_ARCH=${ALPINE_ARCH:-x86_64}
+ALPINE_CDN=${ALPINE_CDN:-http://dl-cdn.alpinelinux.org/alpine}
+CONTAINER_FS_URL=${CONTAINER_FS_URL:-"$ALPINE_CDN/v$ALPINE_RELEASE/releases/$ALPINE_ARCH/alpine-minirootfs-$ALPINE_VERSION-$ALPINE_ARCH.tar.gz"}
+CACHES_VOLUME_NAME=${CACHES_VOLUME_NAME:-trezor_firmware_caches}
```
Do we need these variables configurable? Why not just setting them?

darwin replied:
Personally I'm for optional configurability but all env config vars configurable from outside should have `TREZOR_` prefix to reduce risk of picking something from env accidentally. IMO, this should be standard "hygiene" when writing bash scripts.

tsusanka reviewed (commented):
This fails on my NixOS on Poetry:

```
Installing dependencies from lock file
Package operations: 2 installs, 71 updates, 0 removals
  - Updating pycparser (2.20 /tmp/trezor-firmware/.venv/lib/python3.8/site-packages -> 2.20)

[FileNotFoundError]
[Errno 2] No such file or directory: '/tmp/trezor-firmware/.venv/bin/pip'
```

I am not sure why yet.

---

We are using this mainly for release builds so I am not sure if we do not want to keep it as simple as possible. But looking into it I don't find anything wrong so ccing @prusnak what he thinks.

tsusanka commented:
> This fails on my NixOS on Poetry:

Btw #1330 works just fine.

darwin commented:
My theory would be that your `/tmp` contains something affecting python/pip/poetry which breaks in the presence of mapped `/root/.cache`.

If I understand Nix correctly it uses `/tmp` from host machine as-is and it could in theory contain anything.

I would be interested in two things:

1. seeing if the script works when you start with empty `/tmp`
2. seeing if the script works when you start with clean caches `BLOW_CACHES=1 ./build-docker.sh`

prusnak commented:
Let's focus on reviewing #1330 first - it does not make sense to review this unless #1330 is approved and merged first.

darwin commented:
Ok, I tested how it behaves here on my macOS box and my theory with `/tmp` was wrong. It uses completely clean `/tmp`.

You could add `bash` line after first `poetry install` and explore how `/tmp/trezor-firmware/.venv` looks on your machine (inside Nix).

I could also modify this PR and make caching completely optional behind a flag.

darwin commented:
Rebased and updated. I found a much faster way how to use canonical repo between builds. See the usage of overlayfs inside.

Also I identified the @tsusanka's [problem](https://github.com/trezor/trezor-firmware/pull/1333#pullrequestreview-520458785). In local repo case, I was using rsync to copy local source tree into the container. I tried to reset it into pristine state with `git clean -ffd`. But that wasn't enough, rsync would copy git ignored files but `git clean` would not delete them. So for example `.venv` folder from host machine would ged synced inside, and later confuse the build in the container. The required option is to add `-x` to `git clean` as documented here: https://git-scm.com/docs/git-clean#Documentation/git-clean.txt--x

I have tested this on macOS 11.0 (Big Sur) and under Win10/WSL (ubuntu).

Please remove the cache before testing it on your machine: `docker volume rm trezor_firmware_caches`

matejcik commented:
this seems superseded by https://github.com/trezor/trezor-firmware/pull/2934
